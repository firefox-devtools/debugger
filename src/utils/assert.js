// @flow

function assert(condition: boolean, message: string) {
  if (!condition) {
    throw new Error(`Assertion failure: ${message}`);
  }
}

module.exports = assert;
