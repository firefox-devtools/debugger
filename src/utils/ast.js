/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

import { xor, range } from "lodash";
import { convertToList } from "./pause/pausePoints";

import type { SourceLocation, Position } from "../types";
import type { Symbols } from "../reducers/ast";

import type {
  AstPosition,
  AstLocation,
  PausePoints,
  FunctionDeclaration,
  ClassDeclaration
} from "../workers/parser";

export function findBestMatchExpression(symbols: Symbols, tokenPos: Position) {
  if (symbols.loading) {
    return null;
  }

  const { line, column } = tokenPos;
  const { memberExpressions, identifiers, literals } = symbols;
  const members = memberExpressions.filter(({ computed }) => !computed);

  return []
    .concat(identifiers, members, literals)
    .reduce((found, expression) => {
      const overlaps =
        expression.location.start.line == line &&
        expression.location.start.column <= column &&
        expression.location.end.column >= column;

      if (overlaps) {
        return expression;
      }

      return found;
    }, null);
}

export function findEmptyLines(sourceText: string, pausePoints: PausePoints) {
  if (!pausePoints || !sourceText) {
    return [];
  }

  const pausePointsList = convertToList(pausePoints);

  const breakpoints = pausePointsList.filter(point => point.types.break);
  const breakpointLines = breakpoints.map(point => point.location.line);

  if (!sourceText || breakpointLines.length == 0) {
    return [];
  }

  const lineCount = sourceText.split("\n").length;
  const sourceLines = range(1, lineCount + 1);
  return xor(sourceLines, breakpointLines);
}

export function containsPosition(a: AstLocation, b: AstPosition) {
  const startsBefore =
    a.start.line < b.line ||
    (a.start.line === b.line && a.start.column <= b.column);
  const endsAfter =
    a.end.line > b.line || (a.end.line === b.line && a.end.column >= b.column);

  return startsBefore && endsAfter;
}

function findClosestofSymbol(declarations: any[], location: SourceLocation) {
  if (!declarations) {
    return null;
  }

  return declarations.reduce((found, currNode) => {
    if (
      currNode.name === "anonymous" ||
      !containsPosition(currNode.location, {
        line: location.line,
        column: location.column || 0
      })
    ) {
      return found;
    }

    if (!found) {
      return currNode;
    }

    if (found.location.start.line > currNode.location.start.line) {
      return found;
    }
    if (
      found.location.start.line === currNode.location.start.line &&
      found.location.start.column > currNode.location.start.column
    ) {
      return found;
    }

    return currNode;
  }, null);
}

export function findClosestFunction(
  symbols: ?Symbols,
  location: SourceLocation
): FunctionDeclaration | null {
  if (!symbols || symbols.loading) {
    return null;
  }

  return findClosestofSymbol(symbols.functions, location);
}

export function findClosestClass(
  symbols: Symbols,
  location: SourceLocation
): ClassDeclaration | null {
  if (!symbols || symbols.loading) {
    return null;
  }

  return findClosestofSymbol(symbols.classes, location);
}
