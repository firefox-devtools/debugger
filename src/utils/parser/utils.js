// @flow

const babylon = require("babylon");
const traverse = require("babel-traverse").default;
const t = require("babel-types");
const { isDevelopment } = require("devtools-config");
const toPairs = require("lodash/toPairs");
const get = require("lodash/get");
const isEmpty = require("lodash/isEmpty");

import type { SourceText, Location, Frame, TokenResolution } from "../../types";

const ASTs = new Map();

const symbolDeclarations = new Map();

type ASTLocation = {
  start: {
    line: number,
    column: number
  },
  end: {
    line: number,
    column: number
  }
};

export type SymbolDeclaration = {
  name: string,
  location: ASTLocation
};

export type FormattedSymbolDeclaration = {
  id: string,
  title: string,
  subtitle: string,
  value: string,
  location: ASTLocation
};

export type SymbolDeclarations = {
  functions: Array<FormattedSymbolDeclaration>,
  variables: Array<FormattedSymbolDeclaration>
};

function _parse(code) {
  return babylon.parse(code, {
    sourceType: "module",

    plugins: ["jsx", "flow"]
  });
}

function parse(text: string) {
  let ast;
  if (!text) {
    return;
  }

  try {
    ast = _parse(text);
  } catch (error) {
    if (isDevelopment()) {
      console.warn("parse failed", text);
    }

    ast = {};
  }

  return ast;
}

function getAst(sourceText: SourceText) {
  if (ASTs.has(sourceText.id)) {
    return ASTs.get(sourceText.id);
  }

  const ast = parse(sourceText.text);
  ASTs.set(sourceText.id, ast);
  return ast;
}

function getFunctionName(path) {
  if (path.node.id) {
    return path.node.id.name;
  }

  const parent = path.parent;
  if (parent.type == "ObjectProperty") {
    return parent.key.name;
  }

  if (parent.type == "ObjectExpression" || path.node.type == "ClassMethod") {
    return path.node.key.name;
  }

  if (parent.type == "VariableDeclarator") {
    return parent.id.name;
  }

  if (parent.type == "AssignmentExpression") {
    if (parent.left.type == "MemberExpression") {
      return parent.left.property.name;
    }

    return parent.left.name;
  }

  return "anonymous";
}

function isFunction(path) {
  return t.isFunction(path) ||
    t.isArrowFunctionExpression(path) ||
    t.isObjectMethod(path) ||
    t.isClassMethod(path);
}

function formatSymbol(symbol: SymbolDeclaration): FormattedSymbolDeclaration {
  return {
    id: `${symbol.name}:${symbol.location.start.line}`,
    title: symbol.name,
    subtitle: `:${symbol.location.start.line}`,
    value: symbol.name,
    location: symbol.location
  };
}

function getVariableNames(path) {
  if (t.isObjectProperty(path) && !isFunction(path.node.value)) {
    return [
      formatSymbol({
        name: path.node.key.name,
        location: path.node.loc
      })
    ];
  }

  if (!path.node.declarations) {
    return path.node.params.map(dec =>
      formatSymbol({
        name: dec.name,
        location: dec.loc
      }));
  }

  return path.node.declarations.map(dec =>
    formatSymbol({
      name: dec.id.name,
      location: dec.loc
    }));
}

function isVariable(path) {
  return t.isVariableDeclaration(path) ||
    (isFunction(path) && path.node.params.length) ||
    (t.isObjectProperty(path) && !isFunction(path.node.value));
}

function getMemberExpression(root) {
  function _getMemberExpression(node, expr) {
    if (t.isMemberExpression(node)) {
      expr = [node.property.name].concat(expr);
      return _getMemberExpression(node.object, expr);
    }

    if (t.isThisExpression(node)) {
      return ["this"].concat(expr);
    }
    return [node.name].concat(expr);
  }

  return _getMemberExpression(root, []);
}

function getSymbols(source: SourceText): SymbolDeclarations {
  if (symbolDeclarations.has(source.id)) {
    const symbols = symbolDeclarations.get(source.id);
    if (symbols) {
      return symbols;
    }
  }

  const ast = getAst(source);
  const symbols = { functions: [], variables: [] };

  traverse(ast, {
    enter(path) {
      if (isVariable(path)) {
        symbols.variables.push(...getVariableNames(path));
      }

      if (isFunction(path)) {
        symbols.functions.push(
          formatSymbol({
            name: getFunctionName(path),
            location: path.node.loc
          })
        );
      }

      if (t.isClassDeclaration(path)) {
        symbols.variables.push(
          formatSymbol({
            name: path.node.id.name,
            location: path.node.loc
          })
        );
      }
    }
  });

  symbolDeclarations.set(source.id, symbols);
  return symbols;
}

function resolveExpression(path, token: string, location: Location): ?Object {
  const node = path.node;
  if (
    t.isMemberExpression(node) &&
    node.property.name === token &&
    nodeContainsLocation({ node, location })
  ) {
    const expr = getMemberExpression(node);
    return {
      value: expr.join("."),
      location: node.loc
    };
  }

  return null;
}

function resolveScope(path, location: Location) {
  const node = path.node;
  if (
    (isFunction(path) || t.isProgram(path)) &&
    nodeContainsLocation({ node, location })
  ) {
    return path;
  }
}

// Resolves a token (at location) in the source to determine if it is in scope
// of the given frame and the expression (if any) to which it belongs
function resolveToken(
  source: SourceText,
  token: string,
  location: Location,
  frame: Frame
): ?TokenResolution {
  const ast = getAst(source);
  const scopes = [];
  let expression = null;
  let inScope = false;

  if (isEmpty(ast)) {
    return;
  }

  traverse(ast, {
    enter(path) {
      let scope = null;

      // if we haven't found an expression yet, determine if the token is part
      // of one
      if (!expression) {
        expression = resolveExpression(path, token, location);
      }

      // determine if the current path is a function or program containing the
      // frame
      scope = resolveScope(path, frame.location);
      if (scope) {
        scopes.unshift(scope);
      }
    }
  });

  // determine if the narrowest scope contains the token's location
  inScope = nodeContainsLocation({ node: scopes[0].node, location });

  return {
    expression,
    inScope
  };
}

function nodeContainsLocation({ node, location }) {
  const { start, end } = node.loc;
  const { line, column } = location;

  return !(start.line > line ||
    (start.line === line && start.column > column) ||
    end.line < line ||
    (end.line === line && end.column < column));
}

function getPathClosestToLocation(source: SourceText, location: Location) {
  const ast = getAst(source);
  let pathClosestToLocation = null;

  traverse(ast, {
    enter(path) {
      if (nodeContainsLocation({ node: path.node, location })) {
        pathClosestToLocation = path;
      }
    }
  });

  return pathClosestToLocation;
}

function getVariablesInScope(source: SourceText, location: Location) {
  const path = getPathClosestToLocation(source, location);
  const bindings = get(path, "scope.bindings", {});

  return toPairs(bindings).map(([name, binding]) => ({
    name,
    references: binding.referencePaths
  }));
}

module.exports = {
  getSymbols,
  getPathClosestToLocation,
  getVariablesInScope,
  resolveToken
};
