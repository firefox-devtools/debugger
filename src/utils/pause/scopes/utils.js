/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

import type { Why } from "../../../types";
import type { NamedValue } from "./types";

export function getFramePopVariables(why: Why, path: string): NamedValue[] {
  const vars: Array<NamedValue> = [];

  if (why && why.frameFinished) {
    const frameFinished = why.frameFinished;

    // Always display a `throw` property if present, even if it is falsy.
    if (Object.prototype.hasOwnProperty.call(frameFinished, "throw")) {
      vars.push({
        name: "<exception>",
        path: `${path}/<exception>`,
        contents: { value: frameFinished.throw }
      });
    }

    if (Object.prototype.hasOwnProperty.call(frameFinished, "return")) {
      const returned = frameFinished.return;

      // Do not display undefined. Do display falsy values like 0 and false. The
      // protocol grip for undefined is a JSON object: { type: "undefined" }.
      if (typeof returned !== "object" || returned.type !== "undefined") {
        vars.push({
          name: "<return>",
          path: `${path}/<return>`,
          contents: { value: returned }
        });
      }
    }
  }

  return vars;
}

export function getThisVariable(this_: any, path: string): ?NamedValue {
  if (!this_) {
    return null;
  }

  return {
    name: "<this>",
    path: `${path}/<this>`,
    contents: { value: this_ }
  };
}
