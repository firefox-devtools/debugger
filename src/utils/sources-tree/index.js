/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

/**
 * Utils for Sources Tree Component
 * @module utils/sources-tree
 */

export { addToTree } from "./addToTree";
export { collapseTree } from "./collapseTree";
export { createTree } from "./createTree";
export { formatTree } from "./formatTree";
export { getDirectories } from "./getDirectories";
export { getFilenameFromPath, getURL } from "./getURL";
export { sortEntireTree, sortTree } from "./sortTree";
export { updateTree } from "./updateTree";

export * from "./utils";
