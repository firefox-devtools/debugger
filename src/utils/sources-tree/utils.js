/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

import { parse } from "url";

import type { Node } from "./types";
import type { SourceRecord } from "../../types";
import { isPretty } from "../source";
const IGNORED_URLS = ["debugger eval code", "XStringBundle"];

export function nodeHasChildren(item: Node): boolean {
  return Array.isArray(item.contents);
}

export function isExactUrlMatch(pathPart: string, debuggeeUrl: string) {
  // compare to hostname with an optional 'www.' prefix
  const { host } = parse(debuggeeUrl);
  if (!host) {
    return false;
  }
  return host.replace(/^www\./, "") === pathPart.replace(/^www\./, "");
}

export function isDirectory(url: Object) {
  const parts = url.path.split("/").filter(p => p !== "");

  // Assume that all urls point to files except when they end with '/'
  // Or directory node has children
  return (
    (parts.length === 0 ||
      url.path.slice(-1) === "/" ||
      nodeHasChildren(url)) &&
    url.name != "(index)"
  );
}

export function getExtension(source: Object): string {
  const url = source.get ? source.get("url") : source.url;
  const parsedUrl = parse(url).pathname;
  if (!parsedUrl) {
    return "";
  }
  return parsedUrl.split(".").pop();
}

export function isNotJavaScript(source: Object): boolean {
  return ["css", "svg", "png"].includes(getExtension(source));
}

export function isInvalidUrl(url: Object, source: SourceRecord) {
  return (
    IGNORED_URLS.indexOf(url) != -1 ||
    !source.get("url") ||
    !url.group ||
    isPretty(source) ||
    isNotJavaScript(source)
  );
}

export function partIsFile(index: number, parts: Array<string>, url: Object) {
  const isLastPart = index === parts.length - 1;
  return !isDirectory(url) && isLastPart;
}

export function createNode(
  name: string,
  path: string,
  contents: SourceRecord | Array<Node>
): Node {
  return {
    name,
    path,
    contents
  };
}

export function createParentMap(tree: Node): WeakMap<Node, Node> {
  const map = new WeakMap();

  function _traverse(subtree) {
    if (nodeHasChildren(subtree)) {
      for (const child of subtree.contents) {
        map.set(child, subtree);
        _traverse(child);
      }
    }
  }

  // Don't link each top-level path to the "root" node because the
  // user never sees the root
  tree.contents.forEach(_traverse);
  return map;
}

export function getRelativePath(url: string) {
  const { pathname } = parse(url);
  if (!pathname) {
    return url;
  }
  const path = pathname.split("/");
  path.shift();
  return path.join("/");
}
