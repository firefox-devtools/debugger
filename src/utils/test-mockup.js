/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

/**
 * This file is for use by unit tests for isolated debugger components that do
 * not need to interact with the redux store. When these tests need to construct
 * debugger objects, these interfaces should be used instead of plain object
 * literals.
 */

import type {
  ActorId,
  Breakpoint,
  Expression,
  Frame,
  FrameId,
  Scope,
  JsSource,
  WasmSource,
  Source,
  SourceId,
  Why
} from "../types";

function makeMockSource(
  url: string = "url",
  id: SourceId = "source",
  contentType: string = "text/javascript",
  text: string = ""
): JsSource {
  return {
    id,
    url,
    isBlackBoxed: false,
    isPrettyPrinted: false,
    loadedState: text ? "loaded" : "unloaded",
    relativeUrl: url,
    introductionUrl: null,
    actors: [],
    isWasm: false,
    contentType,
    isExtension: false,
    text
  };
}

function makeMockWasmSource(text: {| binary: Object |}): WasmSource {
  return {
    id: "wasm-source-id",
    url: "url",
    isBlackBoxed: false,
    isPrettyPrinted: false,
    loadedState: "unloaded",
    relativeUrl: "url",
    introductionUrl: null,
    actors: [],
    isWasm: true,
    isExtension: false,
    text
  };
}

function makeMockScope(
  actor: ActorId = "scope-actor",
  type: string = "block",
  parent: ?Scope = null
): Scope {
  return {
    actor,
    parent,
    bindings: {
      arguments: [],
      variables: {}
    },
    object: null,
    function: null,
    type
  };
}

function mockScopeAddVariable(scope: Scope, name: string) {
  if (!scope.bindings) {
    throw new Error("no scope bindings");
  }
  scope.bindings.variables[name] = { value: null };
}

function makeMockBreakpoint(
  source: Source = makeMockSource(),
  line: number = 1,
  column: ?number
): Breakpoint {
  const location = column
    ? { sourceId: source.id, line, column }
    : { sourceId: source.id, line };
  return {
    id: "breakpoint",
    location,
    astLocation: null,
    generatedLocation: location,
    loading: false,
    disabled: false,
    text: "text",
    originalText: "text",
    options: {}
  };
}

function makeMockFrame(
  id: FrameId = "frame",
  source: Source = makeMockSource("url"),
  scope: Scope = makeMockScope(),
  line: number = 4,
  displayName: string = `display-${id}`
): Frame {
  const location = { sourceId: source.id, line };
  return {
    id,
    thread: "FakeThread",
    displayName,
    location,
    generatedLocation: location,
    source,
    scope,
    this: {}
  };
}

function makeMockFrameWithURL(url: string): Frame {
  return makeMockFrame(undefined, makeMockSource(url));
}

function makeWhyNormal(frameReturnValue: any = undefined): Why {
  if (frameReturnValue) {
    return { type: "why-normal", frameFinished: { return: frameReturnValue } };
  }
  return { type: "why-normal" };
}

function makeWhyThrow(frameThrowValue: any): Why {
  return { type: "why-throw", frameFinished: { throw: frameThrowValue } };
}

function makeMockExpression(value: Object): Expression {
  return {
    input: "input",
    value,
    from: "from",
    updating: false
  };
}

export {
  makeMockSource,
  makeMockWasmSource,
  makeMockScope,
  mockScopeAddVariable,
  makeMockBreakpoint,
  makeMockFrame,
  makeMockFrameWithURL,
  makeWhyNormal,
  makeWhyThrow,
  makeMockExpression
};
