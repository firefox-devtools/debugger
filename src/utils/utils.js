// @flow
/* -*- indent-tabs-mode: nil; js-indent-level: 2 -*- */
/* vim: set ft=javascript ts=2 et sw=2 tw=80: */
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/**
 * Utils for utils, by utils
 * @module utils/utils
 */

/**
 * @memberof utils/utils
 * @static
 */
function handleError(err: any) {
  console.log("ERROR: ", err);
}

/**
 * @memberof utils/utils
 * @static
 */
function promisify(context: any, method: any, ...args: any) {
  return new Promise((resolve, reject) => {
    args.push(response => {
      if (response.error) {
        reject(response);
      } else {
        resolve(response);
      }
    });
    method.apply(context, args);
  });
}

/**
 * @memberof utils/utils
 * @static
 */
function endTruncateStr(str: any, size: number) {
  if (str.length > size) {
    return `...${str.slice(str.length - size)}`;
  }
  return str;
}

/**
 * @memberof utils/utils
 * @static
 */
function updateObj<T: Object>(obj: T, fields: $Shape<T>): T {
  return Object.assign({}, obj, fields);
}

/**
 * @memberof utils/utils
 * @static
 */
function throttle(func: any, ms: number) {
  let timeout, _this;
  return function(...args: any) {
    _this = this;
    if (!timeout) {
      timeout = setTimeout(
        () => {
          func.apply(_this, ...args);
          timeout = null;
        },
        ms
      );
    }
  };
}

function waitForMs(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

module.exports = {
  handleError,
  promisify,
  endTruncateStr,
  updateObj,
  throttle,
  waitForMs
};
