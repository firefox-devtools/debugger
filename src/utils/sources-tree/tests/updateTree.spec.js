/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

import { makeMockSource } from "../../../utils/test-mockup";
import { updateTree, createTree } from "../index";

type RawSource = {| url: string, id: string |};

function createSourcesMap(sources: RawSource[]) {
  const sourcesMap = sources.reduce((map, source) => {
    map[source.id] = makeMockSource(source.url, source.id);
    return map;
  }, {});

  return sourcesMap;
}

function formatTree(tree) {
  return JSON.stringify(tree.uncollapsedTree, null, 2);
}

const sources = [
  {
    id: "server1.conn13.child1/39",
    url: "https://davidwalsh.name/"
  },
  {
    id: "server1.conn13.child1/37",
    url: "https://davidwalsh.name/source1.js"
  },
  {
    id: "server1.conn13.child1/40",
    url: "https://davidwalsh.name/source2.js"
  }
];

const debuggeeUrl = "blah";

describe("calls updateTree.js", () => {
  it("adds one source", () => {
    const prevSources = createSourcesMap([sources[0]]);

    const { sourceTree, uncollapsedTree } = createTree({
      debuggeeUrl,
      sources: prevSources,
      projectRoot: ""
    });

    const newTree = updateTree({
      debuggeeUrl,
      prevSources,
      newSources: createSourcesMap([sources[0], sources[1]]),
      uncollapsedTree,
      sourceTree,
      projectRoot: ""
    });

    expect(formatTree(newTree)).toMatchSnapshot();
  });

  it("adds two sources", () => {
    const prevSources = createSourcesMap([sources[0]]);

    const { sourceTree, uncollapsedTree } = createTree({
      debuggeeUrl,
      sources: prevSources,
      projectRoot: ""
    });

    const newTree = updateTree({
      debuggeeUrl,
      prevSources,
      newSources: createSourcesMap([sources[0], sources[1], sources[2]]),
      uncollapsedTree,
      sourceTree,
      projectRoot: ""
    });

    expect(formatTree(newTree)).toMatchSnapshot();
  });

  // NOTE: we currently only add sources to the tree and clear the tree
  // on navigate.
  it("shows all the sources", () => {
    const prevSources = createSourcesMap([sources[0]]);

    const { sourceTree, uncollapsedTree } = createTree({
      debuggeeUrl,
      sources: prevSources,
      projectRoot: ""
    });

    const newTree = updateTree({
      debuggeeUrl,
      prevSources,
      newSources: createSourcesMap([sources[1]]),
      uncollapsedTree,
      sourceTree,
      projectRoot: ""
    });

    expect(formatTree(newTree)).toMatchSnapshot();
  });
});
