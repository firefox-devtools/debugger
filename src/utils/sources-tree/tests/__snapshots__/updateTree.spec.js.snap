// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calls updateTree.js adds one source 1`] = `
"{
  \\"type\\": \\"directory\\",
  \\"name\\": \\"root\\",
  \\"path\\": \\"\\",
  \\"contents\\": [
    {
      \\"type\\": \\"directory\\",
      \\"name\\": \\"davidwalsh.name\\",
      \\"path\\": \\"davidwalsh.name\\",
      \\"contents\\": [
        {
          \\"type\\": \\"source\\",
          \\"name\\": \\"(index)\\",
          \\"path\\": \\"https://davidwalsh.name/\\",
          \\"contents\\": {
            \\"id\\": \\"server1.conn13.child1/39\\",
            \\"url\\": \\"https://davidwalsh.name/\\",
            \\"isBlackBoxed\\": false,
            \\"isPrettyPrinted\\": false,
            \\"loadedState\\": \\"unloaded\\",
            \\"relativeUrl\\": \\"https://davidwalsh.name/\\",
            \\"introductionUrl\\": null,
            \\"actors\\": [],
            \\"isWasm\\": false,
            \\"contentType\\": \\"text/javascript\\",
            \\"text\\": \\"\\"
          }
        },
        {
          \\"type\\": \\"source\\",
          \\"name\\": \\"source1.js\\",
          \\"path\\": \\"davidwalsh.name/source1.js\\",
          \\"contents\\": {
            \\"id\\": \\"server1.conn13.child1/37\\",
            \\"url\\": \\"https://davidwalsh.name/source1.js\\",
            \\"isBlackBoxed\\": false,
            \\"isPrettyPrinted\\": false,
            \\"loadedState\\": \\"unloaded\\",
            \\"relativeUrl\\": \\"https://davidwalsh.name/source1.js\\",
            \\"introductionUrl\\": null,
            \\"actors\\": [],
            \\"isWasm\\": false,
            \\"contentType\\": \\"text/javascript\\",
            \\"text\\": \\"\\"
          }
        }
      ]
    }
  ]
}"
`;

exports[`calls updateTree.js adds two sources 1`] = `
"{
  \\"type\\": \\"directory\\",
  \\"name\\": \\"root\\",
  \\"path\\": \\"\\",
  \\"contents\\": [
    {
      \\"type\\": \\"directory\\",
      \\"name\\": \\"davidwalsh.name\\",
      \\"path\\": \\"davidwalsh.name\\",
      \\"contents\\": [
        {
          \\"type\\": \\"source\\",
          \\"name\\": \\"(index)\\",
          \\"path\\": \\"https://davidwalsh.name/\\",
          \\"contents\\": {
            \\"id\\": \\"server1.conn13.child1/39\\",
            \\"url\\": \\"https://davidwalsh.name/\\",
            \\"isBlackBoxed\\": false,
            \\"isPrettyPrinted\\": false,
            \\"loadedState\\": \\"unloaded\\",
            \\"relativeUrl\\": \\"https://davidwalsh.name/\\",
            \\"introductionUrl\\": null,
            \\"actors\\": [],
            \\"isWasm\\": false,
            \\"contentType\\": \\"text/javascript\\",
            \\"text\\": \\"\\"
          }
        },
        {
          \\"type\\": \\"source\\",
          \\"name\\": \\"source1.js\\",
          \\"path\\": \\"davidwalsh.name/source1.js\\",
          \\"contents\\": {
            \\"id\\": \\"server1.conn13.child1/37\\",
            \\"url\\": \\"https://davidwalsh.name/source1.js\\",
            \\"isBlackBoxed\\": false,
            \\"isPrettyPrinted\\": false,
            \\"loadedState\\": \\"unloaded\\",
            \\"relativeUrl\\": \\"https://davidwalsh.name/source1.js\\",
            \\"introductionUrl\\": null,
            \\"actors\\": [],
            \\"isWasm\\": false,
            \\"contentType\\": \\"text/javascript\\",
            \\"text\\": \\"\\"
          }
        },
        {
          \\"type\\": \\"source\\",
          \\"name\\": \\"source2.js\\",
          \\"path\\": \\"davidwalsh.name/source2.js\\",
          \\"contents\\": {
            \\"id\\": \\"server1.conn13.child1/40\\",
            \\"url\\": \\"https://davidwalsh.name/source2.js\\",
            \\"isBlackBoxed\\": false,
            \\"isPrettyPrinted\\": false,
            \\"loadedState\\": \\"unloaded\\",
            \\"relativeUrl\\": \\"https://davidwalsh.name/source2.js\\",
            \\"introductionUrl\\": null,
            \\"actors\\": [],
            \\"isWasm\\": false,
            \\"contentType\\": \\"text/javascript\\",
            \\"text\\": \\"\\"
          }
        }
      ]
    }
  ]
}"
`;

exports[`calls updateTree.js shows all the sources 1`] = `
"{
  \\"type\\": \\"directory\\",
  \\"name\\": \\"root\\",
  \\"path\\": \\"\\",
  \\"contents\\": [
    {
      \\"type\\": \\"directory\\",
      \\"name\\": \\"davidwalsh.name\\",
      \\"path\\": \\"davidwalsh.name\\",
      \\"contents\\": [
        {
          \\"type\\": \\"source\\",
          \\"name\\": \\"(index)\\",
          \\"path\\": \\"https://davidwalsh.name/\\",
          \\"contents\\": {
            \\"id\\": \\"server1.conn13.child1/39\\",
            \\"url\\": \\"https://davidwalsh.name/\\",
            \\"isBlackBoxed\\": false,
            \\"isPrettyPrinted\\": false,
            \\"loadedState\\": \\"unloaded\\",
            \\"relativeUrl\\": \\"https://davidwalsh.name/\\",
            \\"introductionUrl\\": null,
            \\"actors\\": [],
            \\"isWasm\\": false,
            \\"contentType\\": \\"text/javascript\\",
            \\"text\\": \\"\\"
          }
        },
        {
          \\"type\\": \\"source\\",
          \\"name\\": \\"source1.js\\",
          \\"path\\": \\"davidwalsh.name/source1.js\\",
          \\"contents\\": {
            \\"id\\": \\"server1.conn13.child1/37\\",
            \\"url\\": \\"https://davidwalsh.name/source1.js\\",
            \\"isBlackBoxed\\": false,
            \\"isPrettyPrinted\\": false,
            \\"loadedState\\": \\"unloaded\\",
            \\"relativeUrl\\": \\"https://davidwalsh.name/source1.js\\",
            \\"introductionUrl\\": null,
            \\"actors\\": [],
            \\"isWasm\\": false,
            \\"contentType\\": \\"text/javascript\\",
            \\"text\\": \\"\\"
          }
        }
      ]
    }
  ]
}"
`;
